src/Infrastructure/Process/Runner/AbstractRunner.php

```diff
@@ -31,10 +38,11 @@ abstract protected function executableName(): string;
      */
     private $processFactory;
 
-    public function __construct(ExecutableFinderInterface $executableFinder, ProcessFactoryInterface $processFactory)
+    public function __construct(ExecutableFinderInterface $executableFinder, ProcessFactoryInterface $processFactory, ?int $timeout = 60)
```

Comment:
I assumed we would make `$timeout` a parameter of `::run()` where a client could vary it per call rather than having to commit to the same timeout across the board. What's reasoning behind putting it here?

Reply:
The process runners are abstracted away pretty far from the main interface used by the calling code. Our module uses "begin", "stage", "commit", and "cleanup" methods, which is clearly according to the way the stager library is designed. It doesn't directly deal with the processes. The only way we have to influence these things is to inject the timeout in the constructor.

Reply:
I see, that makes sense. Then I think we should add it as a parameter to the `begin()`, `stage()`, `commit()`, and `cleanup()` methods, where it makes sense alongside the (similar) `$activeDir`, `$stagingDir`, and `$callback` parameters that those methods already take and where you _can_ easily override it. Then move this parameter to the `run()` method and just pass the value through like those other ones.

Reply:
Personally, I don't agree. The idea of a process timeout only makes sense if there is an underlying process invocation, which is an implementation detail (i.e., sometimes we'll use rsync to copy files, other times we'll use Symfony). Also, if we don't inject it, then sites with special needs won't be able to configure or override it at all.

Reply:
You would be able to override it--right on the API method invocation:

```php
$this->stager->stage(
    $composerCommand,
    $stagingDir,
    $callback,
    $timeout // <-- here
);
```

Reply:
That's disadvantageous for two reasons I can think of off the top of my head:

1. A _site_ cannot override that, should they need to modify it across the board. If we make it an injected parameter, they have a few different ways to modify it. Or Automatic Updates can expose it as config.
2. We have to remember to change it and fine-tune it everywhere we make these calls. That's going to lead to bugs and trickiness. Better to change it once than need to remember to change it in many places, and to understand in what context we're changing it each time.
3. It's a completely meaningless parameter for certain backends (i.e., Symfony Filesystem). That makes the API less clear and more confusing.

Reply:
Hmm. Those are interesting points. Let's discuss them tomorrow.

Reply:
If you're still resistant, I have another proposition for you: an `AbstractRunner::setTimeout()` method. This would be container-overridable in Drupal, but also allow changing the process timeout dynamically if needed.